contracts/PLCR.sol

```diff
@@ -0,0 +1,398 @@
+
+    uint256 constant public PCT_BASE = 10 ** 18; // 0% = 0; 1% = 10^16; 100% = 10^18
+
+    bytes32 constant public CREATE_VOTE_ROLE = keccak256("CREATE_VOTE_ROLE");
+
+    // TODO!!!
+    enum TimeUnit { Blocks, Seconds }
+
+    struct UserVote {
+        uint256 lockId;
+        uint256 stake;
+        bytes32 secretHash;
+        bool revealed;
+        bool voteOption;
+        bool claimed;
+    }
+
+    struct Vote {
+        uint64 commitEndDate;
+        uint64 revealEndDate;
+        bool result;
+        bool computed;
+        string metadata;
+        mapping(address => UserVote) userVotes;
+        uint256 votesFor;
+        uint256 votesAgainst;
+        uint256 totalVotes;
+        uint256 slashPool;
+        uint256 paidReward;
+    }
+
+    Vote[] votes;
+
+    mapping(address => mapping(uint256 => bool)) usedLocks;
+
+    event NewVote(uint256 voteId);
+    event CommitedVote(uint256 voteId, address voter, uint256 stake, bytes32 secretHash);
+    event RevealedVote(uint256 indexed voteId, address voter, uint256 stake, bool option);
+    event ClaimedTokens(uint256 indexed voteId, address voter);
+
+    /**
+     * @notice Initialize App with Staking app at `_staking`, needed quorum of `(_voteQuorum - _voteQuorum % 10^16) / 10^14`, minority bloc slash of `(_minorityBlocSlash - _minorityBlocSlash % 10^16) / 10^14`, commit duration of `(_commitDuration - _commitDuration % 86400) / 86400` day `_commitDuration >= 172800 ? 's' : ''` and reveal duration of `(_revealDuration - _revealDuration % 86400) / 86400` day `_revealDuration >= 172800 ? 's' : ''`
+     * @param _staking Address of app used for staking and locking tokens
+     * @param _voteQuorum Percentage of cast votes needed to pass a Vote
+     * @param _minorityBlocSlash Percentage of tokens used for voting that will be redistributed from the losing side to the winning one
+     * @param _commitDuration Duration in seconds of the Commit period
+     * @param _revealDuration Duration in seconds of the Reveal period
+     */
+    function initialize(
+        IStaking _staking,
+        uint256 _voteQuorum,
+        uint256 _minorityBlocSlash,
+        uint64 _commitDuration,
+        uint64 _revealDuration
+    )
+        onlyInit
+        external
+    {
+        initialized();
+
+        require(isContract(_staking));
+        require(_voteQuorum <= PCT_BASE);
+        require(_minorityBlocSlash <= PCT_BASE);
+
+        staking = _staking;
+        voteQuorum = _voteQuorum;
+        minorityBlocSlash = _minorityBlocSlash;
+        commitDuration = _commitDuration;
+        revealDuration = _revealDuration;
+    }
+
+    /**
+     * @notice Create a new vote about "`_metadata`"
+     * @param _script EVM script to be executed on approval
+     * @param _metadata Vote metadata
+     * @return voteId Id for newly created vote
+     */
+    function newVote(bytes _script, string _metadata) isInitialized external auth(CREATE_VOTE_ROLE) returns (uint256 voteId) {
```

Comment:
Script is not saved and therefore it is never executed. PLCR should be a forwarder.